### ISSUE TITLE
"Feature Toggles" in Ringo

### ISSUE THREAD

**Issue by toirl** [ringo-framework/ringo#6]
Based on the experience in eFa development we have long running feature developments where it takes several months to make a feature ready for prime time. However the current work state needs to made available to the customer.

Feature Toggles [0] are an alternative to Feature Branches and might become handy in such situations.

To make the use of Feature Toggles easy in Ringo and most important to show the "right" way we need to implement a "API" for this. First idea is to add helper methods to lib/helpers/appinfo.py to give a consitent way a access config variables in the ini file and make this helper method in the mako templates.

Other ideas are welcome.

Todo:
- [ ] Define syntax for configuration variables in ini file. (e.g feature.foo = True)
- [ ] Implement API
- [ ] Write documentation (in administration section, reference from development?)

[0] http://martinfowler.com/bliki/FeatureToggle.html

**Comment by toirl** [comment]
@ThomasJunk are you interested in working on this?

**Comment by toirl** [comment]
Added more information on feature toogles: http://martinfowler.com/articles/feature-toggles.html

**Comment by ThomasJunk** [comment]
Unfortunately I wasn't notified.
Anyways: Yes, that might be an interesting topic to work on :+1: 

A nice topic for our "work on basics" :-)

**Comment by ThomasJunk** [comment]
A suggestion for API:

I would like to see a property on the request-object ("feature"):

example.mako:
 % if request.features.report
  <div class="report">
  </div>
 %endif

example.py:

@viewconfig(routename="oldfeature")
def oldfeat(request):
  if request.features.newversion:
     return render_to_response('templates/new.mako',
                              {'foo':1, 'bar':2},
                              request=request)
  else:
    return render_to_response('templates/old.mako',
                              {'foo':1, 'bar':2},
                              request=request)

**Comment by toirl** [comment]
> I would like to see a property on the request-object ("feature")

Thanks for your input. Using the request-object might be handy as the request is used in various places anyway.  I currently have no idea how to populate the request object with such properties for each request. Maybe a tween or a subscriber can do the job per request.

Where do you plan to configure the feature toggle? In the ini file or directly in the program code?

My concerns on using the request object directly are the following:

1. There are already some helper methods to get certain aspects of the application. Choosing a another way for Feature Toggles carries the risk to become somewhat inconsistent.
2. Even if seems to be handy to use the request object directly I am not sure if this is how the request object meant to be used. Maybe this is a question which can be answered by the pyramid guys. I know  it is already used for certaing things like session, registry, context etc, but I personally would tend to keep the request as small as possible. But I may be wrong.
3. There is a possible need to put extra effort into this to make it error prone. Eg: Handle case that a feature has been removed or renamed from the config which is still referenced in the code. Maybe its ok to raise an exception here. 

Asuming the configuration is stored in the ini file we need request anyway to access the configuration. My proposal using a helper method looks like this:

```
def is_feature_enabled(request, name):
     """Returns true if the variable 'feature.name' is set to true in the ini file. 
         If not defined or anything else than true the function returns false"""
     return request.registry.setting.get("feature.%s" % name) == True
```

This method can be used in the view  and in the mako in the same way as in your proposal.

Anyway if you have a good idea how to implement it I like to see it.

**Comment by ThomasJunk** [comment]
Okay. (Mis-) using the request object directly isn't quite nice, I have to admit, but I like the API.
request.registry.settings seems the pyramid way to go.

I am okay with a "helper" method.

**Comment by toirl** [comment]
Ok :) Don't get me wrong: The request object is actually used for various things already. So this might be a way to go. I simply  don't know if this usage is what the request is meant for.

I am fine with starting with a helper method and having options to improve that later.
Im the meanwhile I'll ask on how the the request object is meant to be used.

**Comment by toirl** [comment]
Populating the request object seems to be ok if done right. Thank your for  your the trigger to ask for it on the mailinglist as this opens a door to a bunch of options to solve similar problems. See https://groups.google.com/forum/#!topic/pylons-devel/WhUP4mC0vrQ for more details.

However for the first implementation I still prefer to go with the helper method. First because to stay consinstent with the current way such things are done ein Ringo and I want to get a little mit more familiar with @reify and add_request_method. Knowing that populating the request object is ok and doen't have impacts on performance is a great source of possible solutions and might to be a topic for some general improvements in Ringo

**Comment by toirl** [comment]
Thomas,

In contrast to my last comment I decided to add a *ringo* namespace to the request and make the FeatureToggle available under request.ringo.feature.foo.

The current work is availabel in the feature/ringo-request branch (5100bff70f77448f8be6417c2acdbf6364952205)

**Comment by toirl** [comment]
This feature is finally shipped with version 1.17.0
